libc-bottom-half/sources/__environ.c

```diff
@@ -0,0 +1,37 @@
+#include <unistd.h>
+#include <wasi/core.h>
+#include <wasi/libc.h>
+#include <wasi/libc-internal.h>
+
+char **__environ = NULL;
+extern __typeof(__environ) _environ __attribute__((weak, alias("__environ")));
+extern __typeof(__environ) environ __attribute__((weak, alias("__environ")));
+
+// This function is referenced by a weak symbol in crt1.c, and we define
+// it here in the same source file as __environ, so that this function is
+// linked in iff environment variable support is used.
+__wasi_errno_t __wasilibc_populate_environ(void) {
+    __wasi_errno_t err;
+
+    /* Get the sizes of the arrays we'll have to create to copy in the environment. */
+    size_t environ_count;
+    size_t environ_buf_size;
+    err = __wasi_environ_sizes_get(&environ_count, &environ_buf_size);
+    if (err != __WASI_ESUCCESS) {
+        return err;
+    }
+
+    /* Allocate memory for the array of pointers, adding null terminator. */
+    __environ = malloc(sizeof(char *) * (environ_count + 1));
```

Comment:
I think it would be good to add int overflow checks here, because a compiled binary could be launched on different execution environments with different lifecycle. E.g. on some sandboxed application this could potentially be a vulnerability.

Reply:
Good catch, thanks! Looking at this more closely, there were also cases where memory could be leaked or `__environ` could be left pointing to uninitialized memory if a failure happened partway through. I've now fixed these.